packages/preferences/src/browser/views/components/preference-string-input.tsx

```diff
@@ -0,0 +1,54 @@
+
+/********************************************************************************
+ * Copyright (C) 2020 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as React from 'react';
+import { Preference } from '../../util/preference-types';
+
+interface PreferenceStringInputProps {
+    preferenceDisplayNode: Preference.NodeWithValueInSingleScope;
+    setPreference(preferenceName: string, preferenceValue: string): void;
+}
+
+export const PreferenceStringInput: React.FC<PreferenceStringInputProps> = ({ preferenceDisplayNode, setPreference }) => {
+    const { id } = preferenceDisplayNode;
+    const { data, value } = preferenceDisplayNode.preference;
+
+    const externalValue = (value !== undefined ? value : data.defaultValue) || '';
+
+    const [currentTimeout, setCurrentTimetout] = React.useState<number>(0);
+    const [currentValue, setCurrentValue] = React.useState<string>(externalValue);
+
+    React.useEffect(() => {
+        setCurrentValue(externalValue);
+    }, [externalValue]);
+
+    return (
+        <input
+            type="text"
+            className="theia-input"
+            value={currentValue}
+            onChange={e => {
+                const { value: newValue } = e.target;
+                clearTimeout(currentTimeout);
+                const newTimeout = setTimeout(() => setPreference(id, newValue), 750);
```

Comment:
Almost 1 sec before you type something and it takes effect? that's a lot, could you try something like between 50-150 instead. Please review other places.

We usually use `lodash.debounce` for such code.

Reply:
Originally, we used debounce in this case, but chose this solution after careful study of VSCode’s behavior. To observe VSCode, open the settings UI and a `settings.json` next to eachother, and begin typing (slowly or haltingly) in a string input field in the UI. Notice that the file is only updated after you have stopped typing for quite a while.

The problem that this solves is the following:

There must be two-way communication between the UI and the file. The UI causes updates to the file on user input, and the file causes changes in the UI if the user uses the ‘Reset setting’ action in the menu - this could probably be handled locally to the UI - or if the user modifies the file manually - this really can’t be handled locally. In order to allow the file to modify the UI, the UI is sensitive to `onPreferenceChanged` events, but those have a rather long round-trip from the user changing a preference in the UI to the file being modified and saved and that change registering with the service. This led to the ‘sluggishness’ you remarked on in a previous iteration, because the following events took place:

1. User inputs a value
2. `PreferenceService.set` is called
3. Simultaneously: 
  - the user keeps typing, but `debounce` prevents another call to `.set` immediately
  - The preference is saved with the value from (2) and a `onPreferenceChanged` event fires
4. In response, the UI resets the value to the values from step (2), and this makes it looks like the UI is sluggish - the user sees that not everything they typed has appeared yet.

Given that behavior, it seemed that it was a better tradeoff to delay setting the preference until we could be (reasonably) confident that the user was done with their input, rather than interrupting the input with interleaved events, even if `debounce` made them (relatively) infrequent, and that seems to be the same choice made by VSCode.

Reply:
ok, thank you for taking time to explain it